app/utility/base_planning_svc.py

```diff
@@ -49,21 +49,22 @@ async def add_test_variants(self, links, agent, facts=(), rules=()):
             decoded_test = agent.replace(link.command, file_svc=self.get_service('file_svc'))
             variables = re.findall(self.re_variable, decoded_test)
             if variables:
-                relevant_facts = await self._build_relevant_facts(variables, facts)
-                good_facts = await RuleSet(rules=rules).apply_rules(facts=relevant_facts[0])
-                valid_facts = await self._trim_by_limit(decoded_test, good_facts)
-                for combo in list(itertools.product(*valid_facts)):
-                    try:
-                        copy_test = copy.copy(decoded_test)
-                        copy_link = copy.deepcopy(link)
-                        variant, score, used = await self._build_single_test_variant(copy_test, combo, link.ability.executor)
-                        copy_link.command = self.encode_string(variant)
-                        copy_link.score = score
-                        copy_link.used.extend(used)
-                        copy_link.apply_id(agent.host)
-                        link_variants.append(copy_link)
-                    except Exception as ex:
-                        logging.error('Could not create test variant: %s.\nLink=%s' % (ex, link.__dict__))
+                relevant_facts = await self._build_relevant_facts([x for x in variables if '_' not in x], facts)
+                if all([x for x in relevant_facts]):
```

Comment:
Can you elaborate on  this? I think relevant_facts should already be a list. Do we get anything extra by building a new list with the comprehension?

What are the situations where all will return false?

Reply:
It would be nice if we could avoid the extra nesting for the added conditional too, but understand it might be tricky to restructure things around it.   

If the conditional in this form (all + list comprehension), consider adding a brief comment explaining it, to keep something else from breaking something while trying to streamline things :)

Reply:
Yeah, this is a bit ugly, but it was the most straightforward way to prune out situations where there aren't enough matching variables (slight speedup). We're not building a new list, we're making sure that non of the individual options for necessary facts aren't empty, which would invalidate the entire setup (and throw an exception, which does get caught, but still). If that makes sense to you, I'm down to put a comment in explaining what's going on.

Reply:
discussion resolved by 98bd3ff